### PR TITLE
[dynamo, functorch] trace through FunctionalTensor

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126936

Fixes https://github.com/pytorch/pytorch/issues/126236.

Changes:
- `FakeTensorConverter` checks for the `FunctionalTensor(to_functional_tensor(...))` pattern
- `FunctionalTensor` needs to be unwrapped to check for `_is_functional_tensor`
- `FunctionalStorageImpl::frozen_` needs to be preserved in `FakeTensorConverter.from_real_tensor`
- Don't pop `FunctionalMode` from the dispatcher stack when tracing with dynamo

Repro:
```python
import torch
from torch._subclasses.fake_tensor import FakeTensorMode
from torch._subclasses.functional_tensor import FunctionalTensor, FunctionalTensorMode

def f(a, tmp):
    a_view = a.view(-1)
    with torch.no_grad():
        a.set_(tmp)
        a_view.mul_(2)
    return a + tmp

fake_mode = FakeTensorMode()

inp = torch.ones(3, 3, requires_grad=True)
inp = fake_mode.from_tensor(inp, static_shapes=True)

tmp = torch.ones(3, 3, requires_grad=True)
tmp = fake_mode.from_tensor(tmp, static_shapes=True)

with FunctionalTensorMode():
    inp = FunctionalTensor.to_functional(inp)
    tmp = FunctionalTensor.to_functional(tmp)

    print(inp, tmp)
    opt_f = torch.compile(f, backend="eager")
    try:
        opt_f(inp, tmp)
    finally:
        assert torch.is_grad_enabled()
```

Simple script to check for Meta/FakeTensor conversion:
```python
import torch

from torch._subclasses.fake_tensor import FakeTensorMode, FakeTensorConverter
from torch._subclasses.functional_tensor import FunctionalTensor, FunctionalTensorMode
from torch._subclasses.meta_utils import MetaTensorDescriber, MetaConverter

desc = MetaTensorDescriber()
conv = MetaConverter()
conv2 = FakeTensorConverter()

fake_mode = FakeTensorMode()
inp = torch.ones(3, 3, requires_grad=True)
inp = fake_mode.from_tensor(inp, static_shapes=True)
with FunctionalTensorMode():
    inp = FunctionalTensor.to_functional(inp)
    d = desc.describe_tensor(inp)
    # meta = conv.meta_tensor(d)
    fake = conv2.from_real_tensor(fake_mode, inp)
    print(inp)
    print(fake)
```

cc @zou3519 @Chillee @samdow @kshitij12345 @janeyx99 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @bdhirsh

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -320,9 +327,14 @@ def mk_fake_tensor(make_meta_t):
         )
         if out is NotImplemented:
             raise UnsupportedFakeTensorException("meta converter nyi")
+        if isinstance(t, torch._subclasses.functional_tensor.FunctionalTensor):
+            # breakpoint()
+            out = torch._subclasses.functional_tensor.FunctionalTensor(torch._to_functional_tensor(out))
```
Comment (bdhirsh) [inline torch/_subclasses/fake_tensor.py:332]: instead of hardcoding the handling here, I think it might make more sense to handle it where we handle fakeifying all of the other custom tensor types:

(1) in `describe_tensor()`, add a piece of state saying if our tensor is a `FunctionalTensor`: https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/meta_utils.py#L203

(2) add handling in `def meta_tensor()` that branches on that state, and recursively handles `t.elem`

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -136,6 +136,8 @@ def unset_fake_temporarily():
 def is_fake(x):
     if isinstance(x, FakeTensor):
         return True
+    if isinstance(x, torch._subclasses.functional_tensor.FunctionalTensor):
+        x = x.elem
```
Comment (bdhirsh) [inline torch/_subclasses/fake_tensor.py:140]: nit: recurse here for consistency? (`return is_fake(x.elem)`)

**Comment by bdhirsh** [comment]
Main comment is around consistent handling for fakeifying the `FunctionalTensor` properly but otherwise looks good - thanks for finding all the spots!

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (williamwen42) [reply torch/_subclasses/fake_tensor.py]: Currently made the `FunctionalTensor` case recurse back into `from_real_tensor` since `MetaConverter` doesn't handle `FakeTensor`s and the pattern we need to handle is `FunctionalTensor(_to_functional_tensor(FakeTensor(...)))`.

**Comment by zou3519** [comment]
Tbh we need to solve https://github.com/pytorch/pytorch/issues/105929 first because a FunctionalTensorMode is active here. So... it might be a while before this actually works.

**Comment by zou3519** [comment]
Notes: (@ydwu4 @bdhirsh @williamwen42)
- FunctionalTensor differs from eager mode in that it can error out in a very specific case (set_ + data mutation)
- Does this matter? Users aren't going to use FunctionalTensor directly, the way you get here is via torch.cond. But also users generally do not do this (set_ + data mutation)
- To fix this in general, we need to support TorchDispatchMode and also probably put a tensor_flatten/unflatten on FunctionalTensor (and also need to fakeify C++ FunctionalTensorWrapper correctly). Alternatively cond can try functionalizating in a different way (just try to avoid this situation)
- TL;DR: we're OK on punting this to the future because we think it's unlikely to matter in practice at the moment

Followups:
- Yidi to try to construct a test case where this fails for cond and put up an issue for that for tracking. (or maybe it's not a problem, we'll see)
- we should probably file an issue for this this weird thing where FunctionalTensor is fakeified into a regular Tensor
- we should fix the general issue sometime
